pyro/contrib/gp/util.py

```diff
@@ -1,139 +1,246 @@
 from __future__ import absolute_import, division, print_function
 
+from collections import OrderedDict
+
 import torch
 import torch.nn as nn
+from torch.distributions import biject_to, constraints, transform_to
+from torch.nn import Parameter
 
 import pyro
 import pyro.distributions as dist
-from pyro.params import param_with_module_name
+from pyro.contrib import autoname
+from pyro.infer import TraceMeanField_ELBO
+from pyro.infer.util import torch_backward, torch_item
 
 
 class Parameterized(nn.Module):
     """
     Base class for other modules in Gaussin Process module.
 
-    Parameters of this object can be set priors, set constraints, or fixed to a
-    specific value.
+    Parameters of this object can be set constraints, set priors. This is achieved
+    by moving parameter value to a buffer store and creating "root" parameters which
+    are used to generate that parameter's value. For example, if we set a contraint
+    to a parameter, an "unconstrained" parameter will be created.
+
+    By default, when we set a prior to a parameter, an auto Delta guide will be created.
+    We can use the method :meth:`autoguide` to setup other auto guides.
+
+    To fix a parameter to a specific value, it is enough to turn off its "root"
+    parameters' ``requires_grad`` flags.
 
-    By default, data of a parameter is a float :class:`torch.Tensor` (unless we use
-    :func:`torch.set_default_tensor_type` to change default tensor type). To cast these
-    parameters to a correct data type or GPU device, we can call methods such as
+    Note that by default, data of a parameter is a float :class:`torch.Tensor` (unless we
+    use :func:`torch.set_default_tensor_type` to change default tensor type). To cast
+    these parameters to a correct data type or GPU device, we can call methods such as
     :meth:`~torch.nn.Module.double` or :meth:`~torch.nn.Module.cuda`. See
     :class:`torch.nn.Module` for more information.
-
-    :param str name: Name of this object.
     """
-    def __init__(self, name=None):
+    def __init__(self):
         super(Parameterized, self).__init__()
-        self._priors = {}
-        self._constraints = {}
-        self._fixed_params = {}
-        self._registered_params = {}
+        self._constraints = OrderedDict()
+        self._priors = OrderedDict()
+        self._guides = OrderedDict()
+        self._mode = None
 
-        self.name = name
+    def __call__(self, *args, **kwargs):
+        with autoname.name_count():
+            return super(Parameterized, self).__call__(*args, **kwargs)
```

Comment:
Thanks to `autoname.name_count`, if we set priors to variances of the subkernels of `Sum(kern1, kern2)` where both of them are RBF, then latent name of `variance` will be `RBF__1/variance`, `RBF__2/variance` respectively.